File: Programmers/연습문제/Level 2/조이스틱/Solution.py
def solution(name):
    def count_stick(char):
        return min(ord(char) - 65, 90 - ord(char) + 1)

    count_updown = sum(count_stick(char) for char in name if char != 'A')

    count_leftright = len(name) - 1
    for idx in range(len(name)):
        next_idx = idx + 1
        while next_idx < len(name) and name[next_idx] == 'A':
            next_idx += 1
        distance = min(idx, len(name) - next_idx)
        count_leftright = min(count_leftright, idx + len(name) - next_idx + distance)

    return count_updown + count_leftright


if __name__ == "__main__":
    name = "BABBBB"
    print(solution(name))

File: Programmers/연습문제/Level 2/조이스틱/Solution.swift
import Foundation

func solution(_ name:String) -> Int {
    let name = name.utf8.map{ min(Int($0) - 65, 90 - Int($0) + 1) }

    let count_updown = name.reduce(0, +)
    
    var count_leftright = name.count - 1
    let start_idx = 0
    for right_limit_idx in 0..<name.count {
        var left_limit_idx = right_limit_idx + 1
        while left_limit_idx < name.count && name[left_limit_idx] == 0 {
            left_limit_idx += 1
        }
        let min_plus_distance = min(right_limit_idx - start_idx, name.count - left_limit_idx)
        count_leftright = min(count_leftright, (right_limit_idx - start_idx) + (name.count - left_limit_idx) + min_plus_distance)
    }
    
    return count_updown + count_leftright
}


// TC
let name = "JAAAN"
print(solution(name)) // 23
